### PR TITLE
[lexical][lexical-table] Bug Fix: Delete empty paragraph before table

### LINKED ISSUE #8075: Bug: Delete key does not delete empty line before a table
Delete key does not delete empty line before a table

Lexical version: 0.39.0

## Steps To Reproduce

1. Create a content where the first line is empty, followed by a table
2. Place the cursor on the empty first line
3. Press the `Delete` key

Link to code example:
https://playground.lexical.dev
Video:
https://github.com/user-attachments/assets/c1f44ea9-0f94-46f1-bd81-b0d470106943

## The current behavior
The empty line remains unchanged

## The expected behavior
The empty line should be deleted and the table should move up to become the first element in the editor

### PR DESCRIPTION
**Pull request by Rohithmatham12**
## Description

Forward delete from an empty paragraph immediately before a table was swallowed by the shadow-root protection in `RangeSelection.forwardDeletion`. That guard correctly prevents deleting the table itself, but it also returned before the empty paragraph could be removed.

This change preserves the protected table behavior while removing the empty anchor paragraph and moving the selection to the table boundary.

Closes #8075

### Before

Forward delete from an empty paragraph directly before a table left the empty paragraph in place, so the Delete key appeared to do nothing.

### After

Forward delete removes the empty paragraph and keeps the table protected. Added regression coverage for the paragraph-before-table case and for the shared shadow-root path.

Automated tests:

- `node node_modules/vitest/vitest.mjs --project unit packages/lexical-table/src/__tests__/unit/LexicalTableSelection.test.tsx --run`
- `node node_modules/vitest/vitest.mjs --project unit packages/lexical/src/__tests__/unit/LexicalSelection.test.ts --run`
- `node node_modules/prettier/bin/prettier.cjs --check packages/lexical/src/LexicalSelection.ts packages/lexical/src/__tests__/unit/LexicalSelection.test.ts packages/lexical-table/src/__tests__/unit/LexicalTableSelection.test.tsx`
- `node node_modules/eslint/bin/eslint.js packages/lexical/src/LexicalSelection.ts packages/lexical/src/__tests__/unit/LexicalSelection.test.ts packages/lexical-table/src/__tests__/unit/LexicalTableSelection.test.tsx`

### CODE REVIEW

**Review by potatowagon — COMMENTED** [review]
Reviewed by Navi (Tater Thoughts Bobblehead) on behalf of @potatowagon.

**Assessment: LGTM** ✅

This is a clean, well-scoped fix for #8075 (forward delete swallowed by shadow-root guard when an empty paragraph precedes a table).

**What I checked:**
- **Logic correctness**: The new guard only fires when `anchor.type === "element"` AND `anchorNode.isEmpty()` — it won't affect paragraphs with content. The shadow-root protection for the table itself is preserved (the `return true` still fires after removing the empty node).
- **Caret handling**: `$normalizeCaret($getChildCaret(nextSibling, 'next'))` correctly positions selection at the start of the table boundary. `$getCaretRange(caret, caret)` creates a collapsed selection, which is the expected state after forward-deleting an empty paragraph into a table.
- **Edge cases**: The `anchorNode.remove()` is safe because we've confirmed it's empty. No risk of data loss.
- **Test coverage**: The regression test properly reproduces the exact scenario (empty paragraph → table, forward delete from paragraph) and asserts the paragraph is gone while the table remains.
- **www backwards compatibility**: No API changes, no signature changes, no export changes. This is a pure behavioral fix for a previously-broken interaction. Safe for www consumers (MLCComposer, XDSRichText, EPS).

**Minor note**: CLA is unsigned (likely first-time contributor) — that'll need to be resolved before merge.

Ready to approve once CLA is signed and full CI matrix runs.

**Comment by etrepum** [comment]
Title and description doesn't match the [pull request template](https://raw.githubusercontent.com/facebook/lexical/refs/heads/main/.github/pull_request_template.md)

**Comment by mayrang** [comment]
Looks good overall — a few suggestions inline on test coverage, a minor convention thing, and an alternative approach that avoids duplicating the existing backward-delete block in `deleteCharacter`.

File: packages/lexical-table/src/__tests__/unit/LexicalTableSelection.test.tsx [hunk 1]
```diff
@@ -166,5 +167,27 @@ describe('table selection', () => {
         });
       });
     });
+
+    describe('regression #8075', () => {
```
Comment (mayrang) [inline packages/lexical-table/src/__tests__/unit/LexicalTableSelection.test.tsx:170]: This fix covers all `isShadowRoot()` siblings — `CollapsibleContainerNode`, `LayoutContainerNode`, etc. — not just tables. Worth adding a test with a non-table shadow root to make that visible.

File: packages/lexical-table/src/__tests__/unit/LexicalTableSelection.test.tsx [hunk 2]
```diff
@@ -166,5 +167,27 @@ describe('table selection', () => {
...
+      test('forward delete removes an empty paragraph before a table', () => {
+        testEnv.editor.update(
+          () => {
+            const root = $getRoot();
+            const paragraph = $createParagraphNode();
+            root.clear().append(paragraph, tableNode);
+            paragraph.select();
```
Comment (mayrang) [inline packages/lexical-table/src/__tests__/unit/LexicalTableSelection.test.tsx]: `paragraph.select()` returns `RangeSelection` directly — no need for the `$getSelection()` + `$isRangeSelection` round-trip. Saves an import too.

File: packages/lexical-table/src/__tests__/unit/LexicalTableSelection.test.tsx [hunk 3]
```diff
@@ -166,5 +167,27 @@ describe('table selection', () => {
...
+
+            const selection = $getSelection();
+            if (!$isRangeSelection(selection)) {
+              throw new Error('Expected range selection');
+            }
+            selection.deleteCharacter(false);
+
+            expect(root.getChildren()).toEqual([tableNode]);
```
Comment (mayrang) [inline packages/lexical-table/src/__tests__/unit/LexicalTableSelection.test.tsx:180]: PR description mentions selection moves to the table boundary, but nothing asserts that. Checking `anchor.key` after `deleteCharacter` would nail it down.

File: packages/lexical/src/LexicalSelection.ts [hunk 4]
```diff
@@ -1748,6 +1748,18 @@ export class RangeSelection implements BaseSelection {
         (parent === null ? null : parent.getNextSibling());
 
       if ($isElementNode(nextSibling) && nextSibling.isShadowRoot()) {
```
Comment (mayrang) [inline packages/lexical/src/LexicalSelection.ts]: `deleteCharacter` lines 1807–1824 already handle the same pattern for backward delete into a shadow root. If you invert the guard here instead of adding the new block, empty-paragraph + shadow-root falls through to that existing loop — same result, no duplication. Tested locally, all table and selection tests pass.

I'd lean toward the inversion since it keeps one code path instead of two.

**Comment by Rohithmatham12** [comment]
Addressed the review feedback in `11bc41970`:

- switched `forwardDeletion` to invert the shadow-root guard so empty-anchor shadow-root deletion falls through to the existing shared caret path
- updated the table regression test to use the `RangeSelection` returned by `paragraph.select()`
- added an assertion for the post-delete selection anchor
- added a core selection test with `TestShadowRootNode` to cover non-table shadow roots

Re-ran:

```
node node_modules/vitest/vitest.mjs --project unit packages/lexical-table/src/__tests__/unit/LexicalTableSelection.test.tsx --run
node node_modules/vitest/vitest.mjs --project unit packages/lexical/src/__tests__/unit/LexicalSelection.test.ts --run
node node_modules/prettier/bin/prettier.cjs --check packages/lexical/src/LexicalSelection.ts packages/lexical/src/__tests__/unit/LexicalSelection.test.ts packages/lexical-table/src/__tests__/unit/LexicalTableSelection.test.tsx
node node_modules/eslint/bin/eslint.js packages/lexical/src/LexicalSelection.ts packages/lexical/src/__tests__/unit/LexicalSelection.test.ts packages/lexical-table/src/__tests__/unit/LexicalTableSelection.test.tsx
```

**Comment by etrepum** [comment]
Superseded by #9058

**Comment by Rohithmatham12** [comment]
Thanks for the update. Sounds good, I’ll leave this closed since #9058 supersedes it.
